velox/exec/NestedLoopJoinProbe.h

```diff
@@ -31,30 +32,120 @@ class NestedLoopJoinProbe : public Operator {
   RowVectorPtr getOutput() override;
 
   bool needsInput() const override {
-    return !noMoreInput_ && !input_ && !buildSideEmpty_;
+    return state_ == ProbeOperatorState::kRunning && input_ == nullptr &&
+        !noMoreInput_;
   }
 
+  void noMoreInput() override;
+
   BlockingReason isBlocked(ContinueFuture* future) override;
 
-  bool isFinished() override;
+  bool isFinished() override {
+    return state_ == ProbeOperatorState::kFinish;
+  }
+
+ private:
+  // TODO: maybe consolidate initializeFilter routine across Operators like
+  // HashProbe and MergeJoin
+  void initializeFilter(
+      const core::TypedExprPtr& filter,
+      const RowTypePtr& leftType,
+      const RowTypePtr& rightType);
+
+  bool getBuildData(ContinueFuture* future);
+
+  // Get probe side rows count to process on next call to getCrossProduct()
+  vector_size_t getProbeCnt() const;
+
+  // Get cross product of next 'probeCnt' rows of input_ and all rows of
+  // buildData_.value()[buildIndex].
+  // 'outputType' specifies the type of output.
+  // Projections from input_ and buildData_ to the output are specified by
+  // 'probeProjections' and 'buildProjections', respectively. Caller is
+  // responsible for ensuring all columns in outputType is contained in either
+  // projections.
+  // TODO: consider consolidate the routine of producing cartesian product that
+  // can be reused at MergeJoin::addToOutput
+  RowVectorPtr getCrossProduct(
+      vector_size_t probeCnt,
+      const RowTypePtr& outputType,
+      const std::vector<IdentityProjection>& probeProjections,
+      const std::vector<IdentityProjection>& buildProjections);
+
+  // Evaluates joinCondition against the output of getCrossProduct(probeCnt),
+  // returns the result that passed joinCondition, updates probeMatched_,
+  // buildMatched_ accordingly.
+  RowVectorPtr doMatching(vector_size_t probeCnt);
+
+  // Update probeRow_ and buildIndex_ by advance probeRow_ by probeCnt.
+  // Return true if buildIndex_ is points to the end of buildData_.
+  bool advanceIndex(vector_size_t probeCnt);
+
+  bool buildIndexAtEnd() const {
```

Comment:
```
nit: s/buildIndexAtEnd/hasProbedAllBuildData/
```

Reply:
Changed.
Note that for full/right joins, after probing, buildIndex_ is reset and reused for for emitting mis-matched build side data. This method is also used in this scenario.